### ISSUE TITLE
Distributed TensorFlow scaling efficiency with CUDA_VISIBLE_DEVICES

### ISSUE THREAD

**Issue by leesharma** [tensorflow/tensorflow#39207]
Hello! I'm seeing a strange dip in performance when I run distributed training on a subset of available GPUs using CUDA_VISIBLE_DEVICES and TensorFlow 1.x. I included a case using TensorFlow Docker images and the tf_cnn_benchmarks. Is this expected behavior?

Thank you for your help!

**System information**

- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **No (TF CNN benchmarks + official TF docker images)**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Ubuntu 18.04.3 within container (CentOS 7.3 outside, Singularity 3.5)**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow installed from (source or binary): **official TF Docker container: tensorflow/tensorflow:1.15.2-gpu-py3.sif**
- TensorFlow version (use command below): **1.15.2**
- Python version: **3.6.9** 
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: **10**
- GPU model and memory: **GTX 1080Ti - 11GB x8**

**Describe the current behavior**

On a node with 8 GPUs, distributed scaling efficiency on 4 GPUs is significantly (~15%) worse when setting `CUDA_VISIBLE_DEVICE=<subset of nodes>` compared to keeping CUDA_VISIBLE_DEVICES unset or setting it to all physical GPUs, as seen in the graph below (using the scripts in the next section):

![CUDA scaling with different CUDA_VISIBLE_DEVICES](https://user-images.githubusercontent.com/814638/81127747-341ec500-8f0d-11ea-814c-f44a60c3fa6f.png)

In these runs, I set `CUDA_VISIBLE_DEVICES=0,…,(n-1)` in the dotted case (the TensorFlow default order) and `CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7` in the solid case. According to `nvidia-smi`, the same GPUs are working in both cases.

This pattern is also present for different NVIDIA hardware (seen in 11GB GTX 1080Ti, 11GB RTX 2080Ti, and 32GB V100 nodes), different GPU interconnects (seen in PCIe, NVLink, and DMA-only), different all-reduce strategies (seen in NCCL and xring), and different versions of TensorFlow (seen in 1.15.2, 1.14.0, and 1.13.2). I also get these worse performance results for any subset of nodes (e.g. `CUDA_VISIBLE_DEVICES=1,2,3,4,5,6,7` for `--num_gpus=2`), not just the neat case above.

**Describe the expected behavior**

I'd expect that distributed training on 4 nodes with `CUDA_VISIBLE_DEVICES=0,1,2,3` would have similar performance to the same run with `CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7`.

**Standalone code to reproduce the issue**

Setting up the environment:
```bash
$ singularity pull docker://tensorflow/tensorflow:1.15.2-gpu-py3
$ git clone tensorflow/benchmarks
$ cd benchmarks && git checkout cnn_tf_v1.15_compatible
```

1 GPU:
```bash
$ CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
    singularity exec --nv ../tensorflow_1.15.2-gpu-py3.sif \
    python3 scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
      --num_gpus 1 --tf_random_seed 4321 --device gpu \
      --variable_update replicated --all_reduce_spec nccl \
      --local_parameter_device cpu --nodistortions --gradient_repacking 1 \
      --model resnet50 --optimizer momentum --data_name imagenet --batch_size 64

#=> average total images/sec: ~220

$ CUDA_VISIBLE_DEVICES=0 \
    singularity exec --nv ../tensorflow_1.15.2-gpu-py3.sif \
    python3 scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
      --num_gpus 1 --tf_random_seed 4321 --device gpu \
      --variable_update replicated --all_reduce_spec nccl \
      --local_parameter_device cpu --nodistortions --gradient_repacking 1 \
      --model resnet50 --optimizer momentum --data_name imagenet --batch_size 64

#=> average total images/sec: ~220
```

4 GPUs:
```bash
$ CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
    singularity exec --nv ../tensorflow_1.15.2-gpu-py3.sif \
    python3 scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
      --num_gpus 4 --tf_random_seed 4321 --device gpu \
      --variable_update replicated --all_reduce_spec nccl \
      --local_parameter_device cpu --nodistortions --gradient_repacking 1 \
      --model resnet50 --optimizer momentum --data_name imagenet --batch_size 64

#=> average total images/sec: ~840 (~95% scaling efficiency)

$ CUDA_VISIBLE_DEVICES=0,1,2,3 \   # or any other subset of 4-7 GPUs
    singularity exec --nv ../tensorflow_1.15.2-gpu-py3.sif \
    python3 scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
      --num_gpus 4 --tf_random_seed 4321 --device gpu \
      --variable_update replicated --all_reduce_spec nccl \
      --local_parameter_device cpu --nodistortions --gradient_repacking 1 \
      --model resnet50 --optimizer momentum --data_name imagenet --batch_size 64

#=> average total images/sec: ~700 (~80% scaling efficiency)
```

8 GPUs:
```bash
$ CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
    singularity exec --nv ../tensorflow_1.15.2-gpu-py3.sif \
    python3 scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
      --num_gpus 8 --tf_random_seed 4321 --device gpu \
      --variable_update replicated --all_reduce_spec nccl \
      --local_parameter_device cpu --nodistortions --gradient_repacking 1 \
      --model resnet50 --optimizer momentum --data_name imagenet --batch_size 64

#=> average total images/sec: ~1580 (~90% scaling efficiency)
```

**Other info / logs**

**Comment by byronyi** [comment]
1.15.2 is a past release that only receives security fixes. That means even if we could spot the bug and fix it, it won't be back ported to 1.15, not to mention 1.13/1.14.

I will suggest you to use latest tf-nightly and see if the issue still occurs.

**Comment by leesharma** [comment]
Thank you for your quick response!

I re-ran the tests using tf-nightly, and I found a similar pattern:

![tf-nightly_cuda_scaling](https://user-images.githubusercontent.com/814638/81201422-40e5fc00-8f93-11ea-846a-b5a33869fc78.png)

The run scripts are the same except for the container/branch:

```bash
$ singularity pull docker://tensorflow/tensorflow:nightly-gpu-py3
$ git clone tensorflow/benchmarks
$ cd benchmarks && git checkout cnn_tf_v2.1_compatible  # there isn't a 2.2 branch, but I don't think it should matter for this
```

Running with the tf 2.1.0 container produces the same pattern as well.

**Comment by leesharma** [comment]
To get rid of the extra abstraction layer, I ran this test again without Singularity. It looks like this issue is still present.

Creating a clean environment:
```bash
(base)  $ conda create -n tf2.1 python=3.6 tensorflow-gpu==2.1.0 && conda activate tf2.1
(tf2.1) $ git clone tensorflow/benchmarks
(tf2.1) $ cd benchmarks && git checkout cnn_tf_v2.1_compatible
```

Running the benchmark (using xring this time, which makes the discrepancy more dramatic):
```bash
(tf2.1) $ CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
    python scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py --num_gpus 4 --tf_random_seed 4321 --device gpu --variable_update replicated --all_reduce_spec xring --local_parameter_device cpu --nodistortions --gradient_repacking 1 --model resnet50 --optimizer momentum --data_name imagenet --batch_size 64
#=> average total images/sec: ~800 (~93% scaling efficiency)

(tf2.1) $ CUDA_VISIBLE_DEVICES=0,1,2,3 \   # or any other subset of 4-7 GPUs
    python scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py --num_gpus 4 --tf_random_seed 4321 --device gpu --variable_update replicated --all_reduce_spec xring --local_parameter_device cpu --nodistortions --gradient_repacking 1 --model resnet50 --optimizer momentum --data_name imagenet --batch_size 64
#=> average total images/sec: ~600 (~70% scaling efficiency)
```

![tf_scaling_no_container](https://user-images.githubusercontent.com/814638/81334445-0d2ed300-9074-11ea-95ff-8e22db942752.png)

This was run on Centos 7 using CUDA 10.1 and TensorFlow 2.1.0 on a node with 8x 11GB GTX 1080Ti's.

**Comment by zongweiz** [comment]
@leesharma could you please provide more information about your GPU nodes and how they interconnect together (using nvlink etc.)
nvidia-smi topo -m

We have seen some 4GPU cases has lower performance due to the fact that there are no high-speed nvlink among all these GPUs and gradient allreduce needs to go through slower PCI-e.

**Comment by leesharma** [comment]
I tested three different IC configurations. The results above are for running with the same configuration on the same hardware—`CUDA_VISIBLE_DEVICES` is the only thing that's changing.

The graphs above are from nccl/xring all-reduce with PCIe:

```bash
# 1080Ti system
$ nvidia-smi topo -m
        GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    mlx5_0  CPU Affinity
GPU0     X      PIX     PIX     PIX     NODE    NODE    NODE    NODE    NODE    0-15,32-47
GPU1    PIX      X      PIX     PIX     NODE    NODE    NODE    NODE    NODE    0-15,32-47
GPU2    PIX     PIX      X      PIX     NODE    NODE    NODE    NODE    NODE    0-15,32-47
GPU3    PIX     PIX     PIX      X      NODE    NODE    NODE    NODE    NODE    0-15,32-47
GPU4    NODE    NODE    NODE    NODE     X      PIX     PIX     PIX     PIX     0-15,32-47
GPU5    NODE    NODE    NODE    NODE    PIX      X      PIX     PIX     PIX     0-15,32-47
GPU6    NODE    NODE    NODE    NODE    PIX     PIX      X      PIX     PIX     0-15,32-47
GPU7    NODE    NODE    NODE    NODE    PIX     PIX     PIX      X      PIX     0-15,32-47
mlx5_0  NODE    NODE    NODE    NODE    PIX     PIX     PIX     PIX      X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe switches (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing a single PCIe switch
  NV#  = Connection traversing a bonded set of # NVLinks
```

I also saw the same trend on a V100 node with NVLink:
```bash
# V100 system
$ nvidia-smi topo -m
        GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    mlx5_0  mlx5_1  CPU Affinity
GPU0     X      NV1     NV1     NV2     NV2     SYS     SYS     SYS     NODE    SYS     0-15,32-47
GPU1    NV1      X      NV2     NV1     SYS     NV2     SYS     SYS     NODE    SYS     0-15,32-47
GPU2    NV1     NV2      X      NV2     SYS     SYS     NV1     SYS     PIX     SYS     0-15,32-47
GPU3    NV2     NV1     NV2      X      SYS     SYS     SYS     NV1     PIX     SYS     0-15,32-47
GPU4    NV2     SYS     SYS     SYS      X      NV1     NV1     NV2     SYS     PIX     16-31,48-63
GPU5    SYS     NV2     SYS     SYS     NV1      X      NV2     NV1     SYS     PIX     16-31,48-63
GPU6    SYS     SYS     NV1     SYS     NV1     NV2      X      NV2     SYS     NODE    16-31,48-63
GPU7    SYS     SYS     SYS     NV1     NV2     NV1     NV2      X      SYS     NODE    16-31,48-63
mlx5_0  NODE    NODE    PIX     PIX     SYS     SYS     SYS     SYS      X      SYS
mlx5_1  SYS     SYS     SYS     SYS     PIX     PIX     NODE    NODE    SYS      X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks
```

To rule out IC issues, I also tried training with nccl forcing direct memory access for the all-reduce (`NCCL_P2P_DISABLE=1`) on both systems and saw the same pattern in both cases.

**Comment by zongweiz** [comment]
Looks like for the two systems you provided, for example, for 4GPUs (0,1,2,3), NCCL should be able to find optimal cross-GPU connection to achieve maximum allreduce throughput. The issue may not be related to allreduce.

It would be great if you could get some performance profile (for example, using https://www.tensorflow.org/guide/profiler#collect_performance_data).

@reedwm any thoughts from the tf_cnn_benchmarks.py command line options that may be related to the issues?

**Comment by reedwm** [comment]
I general, I recommend using the [TensorFlow Official Models](https://github.com/tensorflow/models/tree/master/official) instead of tf_cnn_benchmarks, as tf_cnn_benchmarks is unmaintained. In this particular case, I can reproduce using both tf_cnn_benchmarks and the Official Models on the latest nightly, so either works in debugging the issue.

The tf_cnn_benchmarks command line flags are fine. `--local_parameter_device cpu` has no impact since nccl is used, but in general it's better to have the local parameter device be the GPU(s).

With tf_cnn_benchmarks, I don't notice any difference in the profile traces (obtained with `--trace_file`). There is a lot of variance when tracing, so if the performance difference is in the trace, I cannot tell. Sometimes the 4-visible devices trace is faster, and sometimes the 8-visible trace is faster. Despite this, the actual 8-visible run is consistently about 15% faster, as reported in the original post. I also checked the post-optimized graphs with `--partitioned_graph_file_prefix`, and they are identical regardless of CUDA_VISIBLE_DEVICES. I am stumped.

@nluehr, any ideas why training would be slower if you restricted the visible GPUs with CUDA_VISIBLE_DEVICES?

**Comment by leesharma** [comment]
As a minor update, I tried running this on some AMD GPUs, and it looks like the same effect is not present—the performance is nearly identical regardless of HIP_VISIBLE_DEVICES.

**Comment by nluehr** [comment]
@leesharma, thanks for reporting this issue. It appears that the GPU kernel launch latency increases when CUDA_VISIBLE_DEVICES is set to more than 1 but fewer than all GPUs. I have filed a bug against the NVIDIA CUDA driver.

**Comment by byronyi** [comment]
@nluehr Thanks for the update! Does this affect containers as well?

**Comment by nluehr** [comment]
If the CUDA_VISIBLE_DEVICES environment variable is used within the container, it will trigger the regression.
If a subset of GPUs are made available within the container (e.g., with `docker run --gpus 4`), I don't see a regression.

**Comment by nluehr** [comment]
@leesharma The issue should be resolved in the 450.51 beta driver which is now available [here](https://www.nvidia.com/Download/Find.aspx?lang=en-us).

**Comment by reedwm** [comment]
I have confirmed this issue does not occur on 450.51, so I'm closing this issue.
